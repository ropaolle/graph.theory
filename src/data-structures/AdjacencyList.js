import Graph from './Graph.js';
import BinaryTree from './BinaryTree.js';
import Tree from './Tree.js';

export default class AdjacencyList {
  constructor(adjacencyObject, type = 'adjacencyList') {
    switch (type) {
      case 'adjacencySet':
        this.list = AdjacencyList.fromSet(adjacencyObject);
        break;
      case 'adjacencyMatrix':
        this.list = AdjacencyList.fromMatrix(adjacencyObject);
        break;
      default:
        this.list = AdjacencyList.fromList(adjacencyObject);
    }
  }

  get nodes() {
    let nodes = new Set();

    for (const [node, neighbors] of Object.entries(this.list)) {
      nodes.add(node);
      for (const element of neighbors) {
        nodes.add(Array.isArray(element) ? element[0] : element);
      }
    }

    return Array.from(nodes);
  }

  get nodeCount() {
    return this.nodes.length;
  }

  get edgeCount() {
    let count = 0;
    for (const neighbors of Object.values(this.list)) {
      count += neighbors.length;
    }

    return count;
  }

  indegree(key) {
    let degree = 0;
    for (const neighbors of Object.values(this.list)) {
      for (const neighbor of neighbors) {
        if (neighbor === key) degree++;
      }
    }
    return degree;
  }

  outdegree(key) {
    if (!this.list[key]) return 0;
    return this.list[key].length;
  }

  degree(key) {
    return this.outdegree(key);
  }

  static fromList(adjacencyListObject) {
    const list = {};

    for (const [node, edges] of Object.entries(adjacencyListObject)) {
      list[node] = [];
      for (const edge of edges) {
        const [destEdge, weight] = edge.split(':');
        list[node].push(weight ? [destEdge, Number(weight)] : destEdge);
      }
    }

    return list;
  }

  static fromSet(adjacencySetObject) {
    const list = {};

    for (const edge of Object.values(adjacencySetObject)) {
      const [sourceEdge, destEdge, weight] = edge.split(':');
      if (list[sourceEdge] === undefined) list[sourceEdge] = [];
      list[sourceEdge].push(weight ? [destEdge, Number(weight)] : destEdge);
    }

    return list;
  }

  static fromMatrix(adjacencyMatrixObject, ignoreNodoToNode = true) {
    const list = {};

    const keys = Object.keys(adjacencyMatrixObject);
    for (const [node, edges] of Object.entries(adjacencyMatrixObject)) {
      list[node] = [];
      for (let i = 0; i < edges.length; i++) {
        if (ignoreNodoToNode && node === keys[i]) continue;
        if (edges[i] === Infinity) continue;
        list[node].push([keys[i], edges[i]]);
      }
    }

    return list;
  }

  static graphToAdjacencyList(graph, weighted = true) {
    let list = {};
    for (const { key } of graph.nodes) {
      if (weighted) {
        list[key] = [];
        for (const [, { a, b, weight }] of graph.edges) {
          if (a === key) list[key].push([b, weight]);
        }
      } else {
        list[key] = graph.adjacent(key);
      }
    }
    return list;
  }

  static toGraph(adjacencyListObject, directed) {
    const graph = new Graph(directed);

    for (const [node, edges] of Object.entries(adjacencyListObject)) {
      if (!graph.findNode(node)) graph.addNode(node);
      for (const edge of edges) {
        if (typeof edge !== 'string') throw new Error('Edge keys must be of type String!');
        const [edgeKey, weight] = edge.split(':');
        if (!graph.findNode(edgeKey)) graph.addNode(edgeKey);
        graph.addEdge(node, edgeKey, weight && Number(weight));
      }
    }

    return graph;
  }

  static toTree(adjacencyListObject, root, binary = false) {
    const tree = binary ? new BinaryTree(root) : new Tree(root);

    const buildTree = (adjacencyList, node, parent) => {
      if (adjacencyList[node]) {
        if (binary && adjacencyList[node].length > 2) {
          throw new Error('Nodes with more than two children is not allowed in a binary tree.');
        }
        for (const child of adjacencyList[node]) {
          if (typeof child !== 'string') throw new Error('Edge keys must be of type String!');
          if (node && node === child) {
            throw new Error(
              'Edges with the same node as the source and target are not allowed (loop).'
            );
          }
          if (child && child === parent) {
            throw new Error('Undirected edges is not allowed (loop to parent).');
          }
          tree.insert(node, child);
          buildTree(adjacencyList, child, node);
        }
      }
      return tree;
    };

    return buildTree(adjacencyListObject, root);
  }

  static toBinaryTree(adjacencyListObject, root) {
    return this.toTree(adjacencyListObject, root, true);
  }

  toString(type = 'graphviz', weight = true) {
    const mermaidCircle = (n) => `${n}((${n}))`;

    let graphStr = '';

    if (type === 'graphviz') {
      graphStr += `digraph MyGraph {\n\n`;
      for (const [node, children] of Object.entries(this.list)) {
        if (weight) {
          for (const [key, value] of children) {
            graphStr += `  ${node} -> ${key} [label="${value}"]\n`;
          }
        } else {
          graphStr += `  ${node} -> {${children.join(', ')}}\n`;
        }
      }
      graphStr += '\n}';
    } else if (type === 'mermaid') {
      graphStr += `\`\`\`mermaid\ngraph LR\n`;
      for (const [node, children] of Object.entries(this.list)) {
        if (weight) {
          for (const [key, value] of children) {
            graphStr += `  ${mermaidCircle(node)} -- ${value} --> ${mermaidCircle(key)}\n`;
          }
        } else {
          graphStr += `  ${node} --> ${children.join(' & ')}\n`;
        }
      }
      graphStr += '```';
    } else if (type === 'd3') {
      let nodes = '  nodes: [\n';
      let links = '  links: [\n';
      for (const [node, children] of Object.entries(this.list)) {
        nodes += `    { id: '${node}' },\n`;
        for (const [key, value] of children) {
          if (weight) {
            links += `    { source: '${node}', target: '${key}', label: '${value}' },\n`;
          } else {
            links += `    { source: '${node}', target: '${key}' },\n`;
          }
        }
      }
      graphStr += `\nconst data = {\n${nodes}  ],\n${links}  ]\n}\n`;
    }

    return graphStr;
  }

  /* istanbul ignore next */ print(type, weight) {
    if (!type) {
      return console.info(this.list);
    }

    if (type === 'raw') {
      const list = {};
      for (const [node, children] of Object.entries(this.list)) {
        list[node] = children.map((child) => (Array.isArray(child) ? child.join(':') : child));
      }
      return console.info(list);
    }

    console.info(this.toString(type, weight));
  }
}
