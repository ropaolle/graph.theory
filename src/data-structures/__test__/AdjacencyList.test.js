import AdjacencyList from '../AdjacencyList.js';

describe('AdjacencyList', () => {
  const listObject = {
    0: ['2', '1'],
    1: ['4', '5'],
    2: ['6', '7'],
    4: ['8'],
    5: ['9'],
    8: ['10', '11'],
  };

  const listWithWeightsObject = {
    A: ['B:1'],
    B: ['C:2'],
    C: ['D:3'],
    D: ['A:3'],
  };

  it('should calculate the tree height correctly', () => {
    const al = new AdjacencyList(listObject);
    expect(al.nodeCount).toBe(11);
    const al2 = new AdjacencyList(listWithWeightsObject);
    expect(al2.nodeCount).toBe(4);
  });

  it('should import list object correctly', () => {
    const al = new AdjacencyList(listObject);
    expect(al.list).toMatchSnapshot();
    expect(AdjacencyList.fromList(listObject)).toMatchSnapshot();
    expect(AdjacencyList.fromList(listWithWeightsObject)).toMatchSnapshot();
  });

  it('should import matrix object correctly', () => {
    const matrixWithWeightsObject = {
      A: [0, 2, 3, 4],
      B: [1, 0, 3, 4],
      C: [1, 2, 0, 4],
      D: [1, 2, 3, 0],
    };
    const al = new AdjacencyList(matrixWithWeightsObject, 'adjacencyMatrix');
    expect(al.list).toMatchSnapshot();
    expect(AdjacencyList.fromMatrix(matrixWithWeightsObject)).toMatchSnapshot();
    const matrixWithWeightsObject2 = {
      A: [0, Infinity, 3, 4],
      B: [1, 0, 3, Infinity],
    };
    expect(AdjacencyList.fromMatrix(matrixWithWeightsObject2)).toMatchSnapshot();
  });

  it('should import set object correctly', () => {
    const setObject = ['A:B', 'B:C', 'C:D', 'D:A', 'D:E'];
    const al = new AdjacencyList(setObject, 'adjacencySet');
    expect(al.list).toMatchSnapshot();
    expect(AdjacencyList.fromSet(setObject)).toMatchSnapshot();
    const setWithWeightsObject = ['A:B:1', 'B:C:2', 'C:D:3', 'D:A:4'];
    const al2 = new AdjacencyList(setWithWeightsObject, 'adjacencySet');
    expect(al2.list).toMatchSnapshot();
    expect(AdjacencyList.fromSet(setWithWeightsObject)).toMatchSnapshot();
  });

  it('should create a Graph object correctly', () => {
    const graph = { A: ['B', 'C', 'D'], B: [1] };
    expect(() => {
      expect(AdjacencyList.toGraph(graph));
    }).toThrowError('Edge keys must be of type String!');
    graph['B'] = 'D';
    expect(AdjacencyList.toGraph(graph)).toMatchSnapshot();
  });

  it('should create a Tree object correctly', () => {
    const graphObject = {
      0: ['1', '2', '5'],
      2: ['3'],
      5: ['4', '6'],
      4: ['9'],
    };
    const tree = AdjacencyList.toTree(graphObject, '0');
    expect(tree.encodeTree()).toBe('(((())())(())())');
    expect(() => {
      AdjacencyList.toTree({ 0: ['0'] }, '0');
    }).toThrowError('Edges with the same node as the source and target are not allowed (loop).');
    expect(() => {
      AdjacencyList.toTree({ 0: ['1'], 1: ['0'] }, '0');
    }).toThrowError('Undirected edges is not allowed (loop to parent).');
    expect(() => {
      AdjacencyList.toTree({ 0: [1], 1: [0] }, '0');
    }).toThrowError('Edge keys must be of type String!');
  });

  it('should create a BinaryTree object correctly', () => {
    const graphObject = {
      0: ['2', '1'],
      1: ['4', '5'],
      2: ['6', '7'],
      4: ['8'],
      5: ['9'],
      8: ['10', '11'],
    };
    const binaryTree = AdjacencyList.toBinaryTree(graphObject, '0');
    expect(binaryTree.toString()).toMatchSnapshot();
    expect(() => {
      AdjacencyList.toBinaryTree({ 0: ['1', '2', '3'] }, '0');
    }).toThrowError('Nodes with more than two children is not allowed in a binary tree.');
  });

  it('should stringify object correctly', () => {
    const al = new AdjacencyList({
      A: ['B:1'],
      B: ['C:2'],
      C: ['D:3'],
      D: ['A:3'],
    });
    const al2 = new AdjacencyList({
      A: ['B'],
      B: ['C'],
      C: ['D'],
      D: ['A'],
    });
    expect(al.toString()).toMatchSnapshot();
    expect(al.toString('', false)).toMatchSnapshot();
    expect(al.toString('graphviz')).toMatchSnapshot();
    expect(al2.toString('graphviz', false)).toMatchSnapshot();
    expect(al.toString('mermaid')).toMatchSnapshot();
    expect(al2.toString('mermaid', false)).toMatchSnapshot();
    expect(al.toString('d3')).toMatchSnapshot();
    expect(al2.toString('d3', false)).toMatchSnapshot();
  });
});
