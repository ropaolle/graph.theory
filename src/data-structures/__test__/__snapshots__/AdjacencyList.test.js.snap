// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdjacencyList should create a BinaryTree object correctly 1`] = `
{
  "0": [
    "2",
    "1",
  ],
  "1": [
    "4",
    "5",
  ],
  "10": [],
  "11": [],
  "2": [
    "6",
    "7",
  ],
  "4": [
    "8",
  ],
  "5": [
    "9",
  ],
  "6": [],
  "7": [],
  "8": [
    "10",
    "11",
  ],
  "9": [],
}
`;

exports[`AdjacencyList should create a Graph object correctly 1`] = `
Graph {
  "directed": true,
  "edges": Map {
    "["A","B"]" => {
      "a": "A",
      "b": "B",
      "weight": undefined,
    },
    "["A","C"]" => {
      "a": "A",
      "b": "C",
      "weight": undefined,
    },
    "["A","D"]" => {
      "a": "A",
      "b": "D",
      "weight": undefined,
    },
    "["B","D"]" => {
      "a": "B",
      "b": "D",
      "weight": undefined,
    },
  },
  "nodes": [
    {
      "key": "A",
      "value": "A",
    },
    {
      "key": "B",
      "value": "B",
    },
    {
      "key": "C",
      "value": "C",
    },
    {
      "key": "D",
      "value": "D",
    },
  ],
}
`;

exports[`AdjacencyList should import list object correctly 1`] = `
{
  "0": [
    "2",
    "1",
  ],
  "1": [
    "4",
    "5",
  ],
  "2": [
    "6",
    "7",
  ],
  "4": [
    "8",
  ],
  "5": [
    "9",
  ],
  "8": [
    "10",
    "11",
  ],
}
`;

exports[`AdjacencyList should import list object correctly 2`] = `
{
  "0": [
    "2",
    "1",
  ],
  "1": [
    "4",
    "5",
  ],
  "2": [
    "6",
    "7",
  ],
  "4": [
    "8",
  ],
  "5": [
    "9",
  ],
  "8": [
    "10",
    "11",
  ],
}
`;

exports[`AdjacencyList should import list object correctly 3`] = `
{
  "A": [
    [
      "B",
      1,
    ],
  ],
  "B": [
    [
      "C",
      2,
    ],
  ],
  "C": [
    [
      "D",
      3,
    ],
  ],
  "D": [
    [
      "A",
      3,
    ],
  ],
}
`;

exports[`AdjacencyList should import matrix object correctly 1`] = `
{
  "A": [
    [
      "B",
      2,
    ],
    [
      "C",
      3,
    ],
    [
      "D",
      4,
    ],
  ],
  "B": [
    [
      "A",
      1,
    ],
    [
      "C",
      3,
    ],
    [
      "D",
      4,
    ],
  ],
  "C": [
    [
      "A",
      1,
    ],
    [
      "B",
      2,
    ],
    [
      "D",
      4,
    ],
  ],
  "D": [
    [
      "A",
      1,
    ],
    [
      "B",
      2,
    ],
    [
      "C",
      3,
    ],
  ],
}
`;

exports[`AdjacencyList should import matrix object correctly 2`] = `
{
  "A": [
    [
      "B",
      2,
    ],
    [
      "C",
      3,
    ],
    [
      "D",
      4,
    ],
  ],
  "B": [
    [
      "A",
      1,
    ],
    [
      "C",
      3,
    ],
    [
      "D",
      4,
    ],
  ],
  "C": [
    [
      "A",
      1,
    ],
    [
      "B",
      2,
    ],
    [
      "D",
      4,
    ],
  ],
  "D": [
    [
      "A",
      1,
    ],
    [
      "B",
      2,
    ],
    [
      "C",
      3,
    ],
  ],
}
`;

exports[`AdjacencyList should import matrix object correctly 3`] = `
{
  "A": [
    [
      undefined,
      3,
    ],
    [
      undefined,
      4,
    ],
  ],
  "B": [
    [
      "A",
      1,
    ],
    [
      undefined,
      3,
    ],
  ],
}
`;

exports[`AdjacencyList should import set object correctly 1`] = `
{
  "A": [
    "B",
  ],
  "B": [
    "C",
  ],
  "C": [
    "D",
  ],
  "D": [
    "A",
    "E",
  ],
}
`;

exports[`AdjacencyList should import set object correctly 2`] = `
{
  "A": [
    "B",
  ],
  "B": [
    "C",
  ],
  "C": [
    "D",
  ],
  "D": [
    "A",
    "E",
  ],
}
`;

exports[`AdjacencyList should import set object correctly 3`] = `
{
  "A": [
    [
      "B",
      1,
    ],
  ],
  "B": [
    [
      "C",
      2,
    ],
  ],
  "C": [
    [
      "D",
      3,
    ],
  ],
  "D": [
    [
      "A",
      4,
    ],
  ],
}
`;

exports[`AdjacencyList should import set object correctly 4`] = `
{
  "A": [
    [
      "B",
      1,
    ],
  ],
  "B": [
    [
      "C",
      2,
    ],
  ],
  "C": [
    [
      "D",
      3,
    ],
  ],
  "D": [
    [
      "A",
      4,
    ],
  ],
}
`;

exports[`AdjacencyList should stringify object correctly 1`] = `
"digraph MyGraph {

  A -> B [label="1"]
  B -> C [label="2"]
  C -> D [label="3"]
  D -> A [label="3"]

}"
`;

exports[`AdjacencyList should stringify object correctly 2`] = `""`;

exports[`AdjacencyList should stringify object correctly 3`] = `
"digraph MyGraph {

  A -> B [label="1"]
  B -> C [label="2"]
  C -> D [label="3"]
  D -> A [label="3"]

}"
`;

exports[`AdjacencyList should stringify object correctly 4`] = `
"digraph MyGraph {

  A -> {B}
  B -> {C}
  C -> {D}
  D -> {A}

}"
`;

exports[`AdjacencyList should stringify object correctly 5`] = `
"\`\`\`mermaid
graph LR
  A((A)) -- 1 --> B((B))
  B((B)) -- 2 --> C((C))
  C((C)) -- 3 --> D((D))
  D((D)) -- 3 --> A((A))
\`\`\`"
`;

exports[`AdjacencyList should stringify object correctly 6`] = `
"\`\`\`mermaid
graph LR
  A --> B
  B --> C
  C --> D
  D --> A
\`\`\`"
`;

exports[`AdjacencyList should stringify object correctly 7`] = `
"
const data = {
  nodes: [
    { id: 'A' },
    { id: 'B' },
    { id: 'C' },
    { id: 'D' },
  ],
  links: [
    { source: 'A', target: 'B', label: '1' },
    { source: 'B', target: 'C', label: '2' },
    { source: 'C', target: 'D', label: '3' },
    { source: 'D', target: 'A', label: '3' },
  ]
}
"
`;

exports[`AdjacencyList should stringify object correctly 8`] = `
"
const data = {
  nodes: [
    { id: 'A' },
    { id: 'B' },
    { id: 'C' },
    { id: 'D' },
  ],
  links: [
    { source: 'A', target: 'B' },
    { source: 'B', target: 'C' },
    { source: 'C', target: 'D' },
    { source: 'D', target: 'A' },
  ]
}
"
`;
